Add V4 distribution implementation developer guide

ricardoV94 opened the pull request:
This PR adds a developer guide on how to implement (and test) new distributions for PyMC3 `v4`. 

Somethings are still marked as `TODO` and feedback is much appreciated. 

The way distributions are implemented (and tested) will probably evolve over time, and hopefully this document will be easy to update as those things happen.

ricardoV94 committed:
Add V4 distribution implementation developer guide

ricardoV94 commented:
@OriolAbril I need quite some help to figure out how to properly integrate this with the rest of the docs, as well things related to markdown formatting, and cross references! :) Thanks in advance

ricardoV94 commented:
@drabbit17, can you have a look at the section on "Adding tests for the new `RandomVariable`", to see if I missed anything obvious, or said something wrong?

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+for parameter broadcasting and shape inference.
+
+Before creating a new `RandomVariable` make sure that it is not offered in the
+[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions)
```
I think this could be `` {class}`numpy.random.Generator` `` or `` {class}`Numpy library <numpy.random.generator>` `` provided we set up a numpy intershpinx mapping, which will probably come in handy in other places,

michaelosthege replied:
```suggestion
[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions).
```

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+Here is a summary check-list of the steps needed to implement a new distribution.
+Each section will be expanded below:
+
+1. Creating a new `RandomVariable` `op`
```
general question, are there docs for RandomVariable and Distribution base classes and if so, do we want to link there?

ricardoV94 replied:
We should. We have docstrings for the distribution base classes. For the RandomVariable I am not sure if it appears in the aesara docs, but if not we can add something for sure

michaelosthege replied:
I find https://aesara.readthedocs.io/en/latest/extending/graphstructures.html to be one of the most useful chapters in the Aesara docs.

michaelosthege replied:
```suggestion
1. Creating a new `RandomVariable` `Op`
```

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+   `ndim_params` attributes. However, the default method may not work for more
+   complex distributions. In that case, custom `_shape_from_params` (and less
+   probably (?) `_infer_shape`) should also be implemented in the new
+   `RandomVariable` class. One simple example is seen in the [MultinomialRV](https://github.com/aesara-devs/aesara/blob/3a3adaee3b8ceccff6d0208996f117fcd5e5eb6f/aesara/tensor/random/basic.py#L518-L521)
```
It looks like MultinomialRV is not anywhere within Aesara's docs, once it is I think this would be better as an intersphinx link. What I don't know how it can be done is linking directly to the source code instead of link to api page that then has the `[source]` link :thinking:

ricardoV94 replied:
The Dirichlet multinomial will have the same trick, so we can link to that #4758

OriolAbril commented on docs/source/distribution_implementation.md:
````diff
@@ -0,0 +1,452 @@
+For a quick test that your new `RandomVariable` `Op` is working, you can just
+call `eval` on a new instance:
+
+```python
````
you have structured all the code as input and output with outputs within the python highlighted code cell, would it make sense to make the file executable? and either run it every time we build docs or save it as ipynb so output gets stored and we don't need to run every time?

ricardoV94 replied:
I am not sure how functional the pseudo code is :/

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+   `PositiveContinuous`, `UnitContinuous`, `BoundedContinuous`,
+   `CircularContinuous`, which specify default transformations for the variables.
+   If you need to specify a one-time custom transform you can also override the
+   `__new__` method, as is done for the [`Dirichlet`](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/multivariate.py#L399-L401).
```
This could also be `` {class}`pymc3.Dirichlet` ``, but again, this goes to the api docs that will have the `[source]` link so it's two clicks, I'll try to find a better way. 

I don't like references to github lines because they either point to a specific commit and get outdated without anyone realizing so checking the link goes to the right code but that code is no longer in use within the actual library or they point to the main/dev branch and any change in line numbers makes the link point to the right file but wrong section.

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+   `__new__` method, as is done for the [`Dirichlet`](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/multivariate.py#L399-L401).
+1. If a distribution does not have a corresponding `random` implementation, a
+   `RandomVariable` should still be created that raises a `NotImplementedError`.
+   This is the case for the [`Flat` distribution](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/continuous.py#L312-L321).
```
With FlatRV the situation is similar yet a bit different, because FlatRV _in not part of the docs!_. I understand it should not be part of the user facing docs but it looks like it's useful for dev facing docs. Would it be a good idea to have a couple of orphan API pages (therefore not indexed in the API navbar section) so we can link there from the developer guide?

ricardoV94 replied:
Sounds reasonable

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+
+## 5. Documenting the new `Distribution`
+
+TODO: What needs to be done? Adding a new entry to the .rst?
```
At the very least, they should have a docstring written and then be referenced in the respective API page. Depending on the distribution and it's uses it should probably get a notebook example where it's used in a case study or a tutorial if it has significant difference all other distributions.

ricardoV94 commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+## 2. Inheriting from a PyMC3 base `Distribution` class
+
+After implementing the new `RandomVariable` `op`, it's time to make use of it
+in a new PyMC3 `distribution`. PyMC3 works in a very functional way, and the
```
We can try that glossary thing with an entry for `functional programming`

matteo-pallini commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,452 @@
+
+This guide does not attempt to explain the rationale behind the `Distributions`
+current implementation, and details are provided only insofar as they help
+implementing new "standard" distributions.
```
>  This guide does not attempt to explain the rationale behind the `Distributions` current implementation

Would you say that [this PR](https://github.com/pymc-devs/pymc3/pull/4440) is a good reference for anyone interested in that? if so, would it make sense to add it here for anyone interested in better understanding the ratio behind the change in implementation?

ricardoV94 replied:
After having a look, I feel it's too unstructured to be of real help. Besides it was not even merged in the end. This one contains all the changes, but it's gigantic: https://github.com/pymc-devs/pymc3/pull/4500

matteo-pallini commented:
> @drabbit17, can you have a look at the section on "Adding tests for the new RandomVariable"

Looks pretty good to me

OriolAbril commented:
Also forgot to mention that this page will now be orphan, and not discoverable in the docs unless one searches for it. Not sure how but it should be added to the toctree or included within he developer guide

ricardoV94 committed:
Add ref links and documentation section

ricardoV94 commented:
> Also forgot to mention that this page will now be orphan, and not discoverable in the docs unless one searches for it. Not sure how but it should be added to the toctree or included within he developer guide

The developer guide should probably be organized hierarchically, instead of being a single file like it is now, no?

OriolAbril commented:
That would probably be best, so it can also include the architecture doc, style guides...

ricardoV94 commented:
I updated the references to follow the format you mentioned `{class}``import`. I added some terms that would go into the glossary, let me know how I can actually add the descriptions.

ricardoV94 commented:
> That would probably be best, so it can also include the architecture doc, style guides...

I would defer that to a separate PR, and in the meantime just leave this guide hanging.

OriolAbril commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+make use of helper methods such as (TODO).
+
+PyMC3 {`Distributions`}`pymc3.distributions.Distribution` build on top of
+Aesara's {`RandomVariables`}`aesara.tensor.random.op.RandomVariable`, and
```
I now realize I did not share any reference on how intersphinx works. The `{class}` thingy indicates that we are linking to an object of type class. intersphinx docs are quite bad so I tried to document everything and add examples at https://arviz-devs.github.io/arviz/contributing/developer_guide.html#reference-external-libraries. It is still missing the custom text option though:

```
{class}`pymc3.distributions.Distribution`                 -> pymc3.distributions.Distribution
{class}`~pymc3.distribution.Distribution`                 -> Distribution
{class}`Distributions <pymc3.distribution.Distribution>`  -> Distributions
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+This guide provides an overview on how to implement a distribution for version 4
+of PyMC3. It is designed for developers who wish to add a new distribution to
+the library. Users will not be aware of all this complexity and should instead
+make use of helper methods such as (TODO).
```
To reduce diff complexity with markdown files, one can do one line per sentence. Single line breaks don't show up in the rendered output and with text I don't think we need to have a limit on the characters per line?

ricardoV94 replied:
Done

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+current implementation, and details are provided only insofar as they help
+to implement new "standard" distributions.
+
+## 1. Creating a new `RandomVariable` `op`
```
```suggestion
## 1. Creating a new `RandomVariable` `Op`
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+class BlahRV(RandomVariable):
+    name: str = "blah"
+
+    # Provide the number of (output) dimensions for this RV
```
```suggestion
    # Provide the minimum number of (output) dimensions for this RV
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+    dtype: str = "floatX"
+
+    # A pretty text and LaTeX representation for the RV
+    _print_name: Tuple[str, str] = ("blah", "\\operatorname{blah}")
```
Maybe choose a more meaningful name so it becomes clear how capitalization should be for these Text/LaTeX fields?

`MoonshotRV` or something.

ricardoV94 replied:
What would that show differently, underscores?

michaelosthege replied:
I'm asking because the LaTeX operators for `Uniform`, or `Normal` start with capital letters, right? But this example does not...

ricardoV94 replied:
Ahhh... I'll change it then. Missed that one

ricardoV94 replied:
```suggestion
    _print_name: Tuple[str, str] = ("Blah", "\\operatorname{Blah}")
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+        return scipy.stats.blah.rvs(loc, scale, random_state=rng, size=size)
+
+# Create an instance of the `RandomVariable` `Op`...
+blah = BlahRV()
```
The user API is always through things like `pm.Normal`, so what's the purpose of these instances?

ricardoV94 replied:
This is what actually creates the `RandomVariable Op`.

ricardoV94 replied:
Added a more clear comment, not sure if I am saying something incorrect

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+   distributions. Unfortunately there is no general reference documenting how
+   `size` ought to work for multivariate distributions.
+   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
+   to get more context.
```
```suggestion
1. The `size` argument (together with the inputs shape) are the one way to specify
    the dimensions (minus the last `ndim_supp`) at the `RandomVariable` API level.
    At the user API level more options (`shape`, `dims`) are available, but this is taken
    care of by {`class`}`~pymc3.distributions.Distribution`.
   The `rng_fn` will have to take `size` into consideration for correct output.
   `size` is the specification used by `Numpy` and `Scipy` and works like
   PyMC3 `shape` for univariate distributions, but is different for multivariate
   distributions. Unfortunately there is no general reference documenting how
   `size` ought to work for multivariate distributions.
   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
   to get more context.
```

ricardoV94 replied:
Added a slightly different note on this below.

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+
+
+For a quick test that your new `RandomVariable` `Op` is working, you can just
+call `eval` on a new instance:
```
```suggestion
For a quick test that your new `RandomVariable` `Op` is working, you can call
the `Op` with parameters and call `eval()` on the resulting {class}`~aesara.tensor.var.TensorVariable` instance:
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+
+## 2. Inheriting from a PyMC3 base `Distribution` class
+
+After implementing the new `RandomVariable` `op`, it's time to make use of it
```
```suggestion
After implementing the new `RandomVariable` `Op`, it's time to make use of it
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+in a very {functional}`Functional_Programming` way, and the `distribution`
+classes are there mostly to facilitate porting the `v3` code to the new `v4`
+version, and to keep related methods organized together. In practice, they take
+care of:
```
```suggestion
classes are there mostly to facilitate porting the `v3` code to the new `v4`
version, add PyMC3 API features and keep related methods organized together.
In practice, they take
care of:
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+   In this case it will be necessary to provide a standard `initval` by
+   overriding `__new__`.
+1. As mentioned above, `v4` works in a very {functional}`Functional_Programming`
+   way, and all the   information that is needed in the `logp` and `logcdf`
```
```suggestion
   way, and all the information that is needed in the `logp` and `logcdf`
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+1. As mentioned above, `v4` works in a very {functional}`Functional_Programming`
+   way, and all the   information that is needed in the `logp` and `logcdf`
+   methods is expected to be "carried" via the `RandomVariable` inputs. You may
+   pass numerical arguments that are not strictly needed for the `random` method
```
```suggestion
   pass numerical arguments that are not strictly needed for the `rng_fn` method
```
?

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+This is usually needed for Multivariate distributions. You can see an example in
+the {`Dirichlet` test}`pymc3.test.test_random.TestDirichlet`.
+
+### Notes on `check_pymcs_draws_match_reference` test
```
Maybe move this into a docstring right at the test definition?

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+   parameters are randomly tested.
+1. On Github all tests are run twice, under the `aesara.config.floatX` flags of
+   `"float64"` and `"float32"`. However, the reference Python functions will
+   run in a pure "float64" environment, which means the reference and the `Pymc3`
```
```suggestion
   run in a pure "float64" environment, which means the reference and the `PyMC3`
```

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,543 @@
+     import numpy as np
+     import scipy.stats as st
+     import arviz as az
+     plt.style.use('arviz-darkgrid')
```
```suggestion
```
These style changes are unrelated to the code that's being demonstrated. If we consider this important for the documentation we should set the style at docs-build-time.

michaelosthege reviewed (changes requested):
Looks almost complete to me!
Let's address the ToDos and merge it sooner than later.

ricardoV94 committed:
Apply suggestions from code review

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

ricardoV94 committed:
Remove line breaks

ricardoV94 committed:
Fix cross-refs and integrate more suggestions

ricardoV94 committed:
Add docstrings to `BaseTestDistribution`

michaelosthege commented on docs/source/distribution_implementation.md:
```diff
@@ -0,0 +1,425 @@
+
+This guide provides an overview on how to implement a distribution for version 4 of PyMC3.
+It is designed for developers who wish to add a new distribution to the library.
+Users will not be aware of all this complexity and should instead make use of helper methods such as (TODO).
```
Leftover TODO

ricardoV94 replied:
It's still a pymc3 ToDo. This was the role of `Densitydist` but that was deprecated or at least not refactored for V4. Maybe we can link to something like: https://github.com/pymc-devs/pymc-examples/pull/185

zoj613 commented:
I think adding information about extending `Op` for new distributions would be valuable. Aesara contains little to no information about this. The example at https://aesara.readthedocs.io/en/latest/extending/extending_aesara.html#example-op-definition is unhelpful and seems very incomplete, nor does it ever touch on the "optional" methods. Moreover, details about dealing with shapes is completely missing. It has been quite frustrating for me to plough through multiple pages of the docs and still not get it.

ricardoV94 commented:
> I think adding information about extending `Op` for new distributions would be valuable. Aesara contains little to no information about this. The example at https://aesara.readthedocs.io/en/latest/extending/extending_aesara.html#example-op-definition is unhelpful and seems very incomplete, nor does it ever touch on the "optional" methods. Moreover, details about dealing with shapes is completely missing. It has been quite frustrating for me to plough through multiple pages of the docs and still not get it.

Do you want to open a Github issue for this? I think it could be important, but perhaps out of scope for this guide.

ricardoV94 commented:
The TODO may point to https://github.com/pymc-devs/pymc-examples/pull/185 in the future, but that is still an early WIP. I think it's fine to leave it for the time-being. Similarly, the glossary will not work until #4797 is wrapped up. 

I am okay merging this and leaving it as an orphan-page until the "dev-branch" that @martinacantaro has suggested is created. If not, we can leave this as a draft PR. As you guys see fit.

ricardoV94 committed:
Small tweaks
